src/connection/socket-connection.ts

```diff
@@ -0,0 +1,215 @@
+import type { Dwn, GenericMessage } from "@tbd54566975/dwn-sdk-js";
+import { DwnMethodName } from "@tbd54566975/dwn-sdk-js";
+
+import type { WebSocket } from "ws";
+import log from 'loglevel';
+import { v4 as uuidv4 } from 'uuid';
+
+import type { RequestContext } from "../lib/json-rpc-router.js";
+import type { JsonRpcErrorResponse, JsonRpcId, JsonRpcRequest, JsonRpcResponse } from "../lib/json-rpc.js";
+
+import { requestCounter } from "../metrics.js";
+import { jsonRpcApi } from "../json-rpc-api.js";
+import { JsonRpcErrorCodes, createJsonRpcErrorResponse, createJsonRpcSuccessResponse } from "../lib/json-rpc.js";
+import { DwnServerError, DwnServerErrorCode } from "../dwn-error.js";
+
+const HEARTBEAT_INTERVAL = 30_000;
+
+export interface JsonRpcSubscription {
+  /** JSON RPC Id of the Subscription Request */
+  id: JsonRpcId;
+  close: () => Promise<void>;
+}
+
+/**
+ * SocketConnection handles a WebSocket connection to a DWN using JSON RPC.
+ * It also manages references to the long running RPC subscriptions for the connection.
+ */
+export class SocketConnection {
+  private heartbeatInterval: NodeJS.Timer;
+  private subscriptions: Map<JsonRpcId, JsonRpcSubscription> = new Map();
+  private isAlive: boolean;
+
+  constructor(
+    private socket: WebSocket,
+    private dwn: Dwn
+  ){
+    socket.on('message', this.message.bind(this));
+    socket.on('close', this.close.bind(this));
+    socket.on('error', this.error.bind(this));
+    socket.on('pong', this.pong.bind(this));
+
+    // Sometimes connections between client <-> server can get borked in such a way that
+    // leaves both unaware of the borkage. ping messages can be used as a means to verify
+    // that the remote endpoint is still responsive. Server will ping each socket every 30s
+    // if a pong hasn't received from a socket by the next ping, the server will terminate
+    // the socket connection
+    this.isAlive = true;
+    this.heartbeatInterval = setInterval(() => {
+      if (this.isAlive === false) {
+        this.close();
+      }
+      this.isAlive = false;
+      this.socket.ping();
+    }, HEARTBEAT_INTERVAL);
+  }
+
+  /**
+   * Adds a reference for the JSON RPC Subscription to this connection.
+   * Used for cleanup if the connection is closed.
+   */
+  async subscribe(subscription: JsonRpcSubscription): Promise<void> {
```

Comment:
Can we rename to `addSubscription() instead?

Reply:
Was originally that 🤣 so definitely do't mind.